source/SIL.AppBuilder.Portal.Frontend/src/ui/components/inputs/locale-input/-utils/helpers.tsx

```diff
@@ -0,0 +1,158 @@
+import * as React from 'react';
+
+export const hasSearchTerm = (searchTerm: string) => {
+  let lowerSearch = (searchTerm || '').trim().toLocaleLowerCase();
+
+  return (property: string) => {
+    try {
+      return (property || '').toLocaleLowerCase().includes(lowerSearch);
+    } catch (e) {
+      console.error(`searched for ${searchTerm} against value: ${property}`, e);
+    }
+  };
+};
+
+export const getSuggestions = (data: ILanguageInfo[]) => (value) => {
+  const has = hasSearchTerm(value);
+
+  return data.filter(
+    ({ name, nameInLocale, region, tag, localname, regions, names }) =>
+      has(nameInLocale) ||
+      has((names || []).join()) ||
+      has(localname || name) ||
+      has(regions || region) ||
+      has(tag)
+  );
+};
+
+export const getSuggestionValue = (suggestion: ILanguageInfo) => suggestion.tag;
+
+export function findLanguageCode(data: ILanguageInfo[]) {
+  return (value: string) => {
+    const info = data.find((lang) => lang.tag === value);
+
+    return (info && info.tag) || '';
+  };
+}
+
+export function findLanguageInfo(data: ILanguageInfo[]) {
+  return (value: string) => {
+    const info = data.find((lang) => lang.tag === value);
+
+    return info;
+  };
+}
+
+export function highlightIfPresent(term: string, highlight: string) {
+  let termLower = term.toLowerCase();
+  let highLower = highlight.toLowerCase();
+
+  if (!termLower.includes(highLower)) {
+    return term;
+  }
+
+  const beginIdx = termLower.indexOf(highLower);
+  const endIdx = beginIdx + highlight.length - 1;
+  const beginning = term.substring(0, beginIdx);
+  const ending = term.substring(endIdx + 1);
+  const termHighlight = term.slice(beginIdx, endIdx + 1);
+
+  return (
+    <span className='phrase-with-highlight'>
+      {beginning}
+      <span className='highlight'>{termHighlight}</span>
+      {ending}
+    </span>
+  );
+}
+
+// return values:
+// -1 a is less than b
+// 1 a is more than b
+// 0 a and b are the same
+//
+// logic from:
+// https://github.com/sillsdev/libpalaso/blob/master/SIL.WritingSystems/LanguageLookup.cs#L290
+//
+// this is a straight port of the C# code
+//
+/// Sorting the languages for display is tricky: we want the most relevant languages at the
+/// top of the list, so we can't simply sort alphabetically by language name or by language tag,
+/// but need to take both items into account together with the current search string.  Ordering
+/// by relevance is clearly impossible since we'd have to read the user's mind and apply that
+/// knowledge to the data.  But the heuristics we use here may be better than nothing...
+export function sortComparer(searchTerm?: string) {
+  const lowerSearch = `${searchTerm}`.toLocaleLowerCase();
```

Comment:
if `searchTerm` is undefined `lowerSearch` will equal `"undefined"` is this ok?

Reply:
nope! thanks!